### PR TITLE
feat: add Partials class

### PR DESCRIPTION
**Pull request by ImRodry**
**Please describe the changes this PR makes and why it should be merged:**

This PR transforms Partials into a class that extends BitField so that we can have properties like All, which adds all partials into an array, and other methods from the inherited class

**Status and versioning classification:**
- Code changes have been tested against the Discord API, or there are no code changes
- I know how to update typings and have done so, or typings don't need updating
- This PR changes the library's interface (methods or parameters added)

### CODE REVIEW

**Review by ItsJackson — COMMENTED** [review]
There’s a error running the ESLint check, with code style issues in these file(s)

**Review by ItsJackson — APPROVED** [review]

**Comment by vladfrangu** [comment]
And what exactly is the benefit from this? It feels like overcomplicating for no reason 🤔

**Comment by ImRodry** [comment]
Quoting from the pr description:
> so that we can have properties like All, which adds all partials into an array, and other methods from the inherited class

**Comment by almeidx** [comment]
> so that we can have properties like All, which adds all partials into an array

Is `Partials.All` something that is actually wanted though? Perhaps not, for the same reasoning as to why `Intents.ALL` was removed (#6050)

**Comment by ImRodry** [comment]
> > so that we can have properties like All, which adds all partials into an array
> 
> Is `Partials.All` something that is actually wanted though? Perhaps not, for the same reasoning as to why `Intents.ALL` was removed (#6050)

Intents.All wasn't wanted because people are not supposed to have all intents and they should specify the ones they want individually. The same doesn't apply to Partials, where you can have all partials enabled with no issues

**Comment by vladfrangu** [comment]
You can have all intents enabled too and you won't have issues, so I don't get why that matters 👀

**Comment by ImRodry** [comment]
That's not what I meant. You're not supposed to have all intents enabled because you often won't need all of them, but you may need all partials and there's nothing wrong with having them all enabled

**Comment by xHyroM** [comment]
I'm leaning towards deleting Partials.All. It's useless, you rarely need all partials.

**Comment by almeidx** [comment]
> and there's nothing wrong with having them all enabled

I disagree. Let's say a new partial is added and your bot is dependent on events/caches of that structure. It now might receive partial structures and you very likely will need to adapt your code if you were accessing data that can be missing (which could not be missing previously).

And besides that, `Partials.All` doesn't actually tell you *what* partials are enabled, and you'd likely need to go to the documentation to find out what partials exist. As opposed to passing all the partials that you need, which is more verbose and doesn't require an external service (e.g. the documentation) to know what is it you're enabling (as is the case for the intents)

**Comment by ImRodry** [comment]
I assume you’d only enable all partials if you knew what partials there were to begin with. Not only that but new partials are only ever added when new structures are added (for example when scheduled events came out) so you wouldn’t need to adapt your code since you wouldn’t have event listeners for something that just came out

**Comment by monbrey** [comment]
I feel like this change is a little counter-intuitive and doesn't really deliver any benefit over the base enum/constant.

Partials aren't (or weren't) a calculated Bitfield and I don't see the need for them to be. Is this all just to have `Partials.All` added?

**Comment by ImRodry** [comment]
Mainly yes but if we want to move away from strings and partials are now treated as numbers and not strings, it would make sense to treat Partials as a Bitfield IMO

**Comment by vladfrangu** [comment]
So, in a nutshell, this PR...

1. Adds a Partials.All which is not even checked by the rest of the flags (like ADMINISTRATOR is checked for permissions), because adding all the 4 partials we have that you need to an array is..annoying?
2. Changes an enum of Strings -> numbers to a bitfield class with a Flags property (which we moved away from); the enum was consistent with the general move in v14 from strings to numbers as is, and there were no more magic numbers/strings
3. Goes against intents where you're supposed to only use the ones you need

I don't know what the other maintainers think, but to me this sounds like bloat and overcomplicating an already complex concept. Do note that this is *my personal* opinion, and now saying this in the name of the entire team. But I don't see any real benefit from this 😅

**Comment by ImRodry** [comment]
So let me address each one of those points individually
1. Partials.All being an array of numbers is correct (this also made me realize there was a bug which I just fixed) because when you access an enum property it's returned as a number, so all properties of that enum would be an array of all the numbers. Also you have 6, not 4 partial types, and most of the time you do want all of them to make sure you get all the data from events. This would also not be an issue when a new partial type is added because, as mentioned previously, this only happens when a new structure is added, so you'd need to add event listeners for that anyway
2. The flags property is literally the exact same as the current Flags enum, I just copied the code over so it's the exact same
3. I don't understand the comparison with intents because there is an actual downside to using all intents, whereas the same doesn't apply to partials, as mentioned previously

File: packages/discord.js/typings/index.d.ts [hunk 1]
```diff
@@ -4788,7 +4793,7 @@ export interface PartialRoleData extends RoleData {
   id?: Snowflake | number;
 }
 
-export enum Partials {
+export enum PartialsFlags {
```
Comment (almeidx) [inline packages/discord.js/typings/index.d.ts:4796]: Doesn't look like this is exported

Comment (ImRodry) [reply packages/discord.js/typings/index.d.ts:4796]: I’ll just not export this then

File: packages/discord.js/typings/index.d.ts [hunk 2]
```diff
@@ -1705,6 +1705,11 @@ export class OAuth2Guild extends BaseGuild {
   public permissions: Readonly<PermissionsBitField>;
 }
 
+export class Partials extends BitField<keyof typeof PartialsFlags> {
+  public static Flags: typeof PartialsFlags;
+  public static All: number[];
```
Comment (almeidx) [inline packages/discord.js/typings/index.d.ts:1710]:
```suggestion
  public static All: PartialsFlags[];
```

**Review by almeidx — CHANGES_REQUESTED** [review]

**Comment by ImRodry** [comment]
@iCrawl would appreciate an explanation on why this was closed

**Comment by iCrawl** [comment]
Explanations were given above.

**Pull request closed without merging**